master/internal/command/command.go

```diff
@@ -19,124 +20,116 @@ import (
+	command struct {
+		mu sync.Mutex
 
-		taskID:   taskID,
-		taskType: taskType,
-		jobType:  jobType,
-		jobID:    jobID,
+		db *db.PgDB
+		rm rm.ResourceManager
 
-		logCtx: logger.Context{
-			"job-id":    jobID,
-			"task-id":   taskID,
-			"task-type": taskType,
-		},
+		tasks.GenericCommandSpec
 
-		contextDirectory: contextDirectory,
+		registeredTime time.Time
+		taskID         model.TaskID
+		taskType       model.TaskType
+		jobType        model.JobType
+		jobID          model.JobID
+		allocationID   model.AllocationID
+		lastState      task.AllocationState
+		exitStatus     *task.AllocationExited
+		restored       bool
+
+		contextDirectory []byte // Don't rely on this being set outsides of PreStart non restore case.
+
+		logCtx logger.Context
+		syslog *logrus.Entry
 	}
 
-	a, _ := ctx.ActorOf(cmd.taskID, cmd)
-	summaryFut := ctx.Ask(a, getSummary{})
-	if err := summaryFut.Error(); err != nil {
-		return errors.Wrap(err, "failed to create generic command")
+	// CreateGeneric is a request to the commandService to create a generic command.
+	CreateGeneric struct {
+		ContextDirectory []byte
+		Spec             *tasks.GenericCommandSpec
 	}
-	// Sync with the actor, but we don't really need the summary. actor.Ping works too,
-	// but this makes sure it can form some sort of useful response (ping doesn't actually
-	// hit the receive block).
-	summaryFut.Get()
-	return nil
-}
+)
 
-func commandFromSnapshot(
-	ctx *actor.Context,
-	db *db.PgDB,
-	rm rm.ResourceManager,
-	snapshot *CommandSnapshot,
-) *command {
-	taskID := snapshot.TaskID
-	taskType := snapshot.Task.TaskType
-	jobID := snapshot.Task.Job.JobID
-	cmd := &command{
-		db:             db,
-		rm:             rm,
-		registeredTime: snapshot.RegisteredTime,
+// newCommandService creates a new commandService.
+func newCommandService() *commandService {
+	return &commandService{
+		commands: make(map[model.TaskID]*command),
+		syslog:   logrus.WithField("component", "command-service"),
+	}
+}
 
-		GenericCommandSpec: snapshot.GenericCommandSpec,
+// InitCommandService starts the global command service singleton.
+func (cs *commandService) InitCommandService(ctx context.Context, db *db.PgDB, rm rm.ResourceManager) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
 
-		taskID:   taskID,
-		taskType: taskType,
-		jobType:  snapshot.Task.Job.JobType,
-		jobID:    jobID,
+	cs.db = db
+	cs.rm = rm
+}
 
-		logCtx: logger.Context{
-			"job-id":    jobID,
-			"task-id":   taskID,
-			"task-type": taskType,
-		},
+// RestoreAllCommands restores all terminated commands whose end time isn't set.
+func (cs *commandService) RestoreAllCommands(
+	ctx context.Context,
+) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
 
-		restored: true,
+	commands, err := cs.remakeAllCommands(cs.db, cs.rm)
+	if err != nil {
+		cs.syslog.Errorf("failed to restoreAllCommands: %s", err)
 	}
 
-	return cmd
+	for _, cmd := range commands {
+		cs.commands[cmd.taskID] = cmd
+		if err := cmd.startCmd(ctx); err != nil {
```

Comment:
let's just make `commandFromSnapshot` call `startCmd` like `createGeneric..` does. more of "things should be constructed correctly by their constructor, not have methods users need to call to make them correct".